### ISSUE TITLE
performance issue after .NET 4.6 installed

### ISSUE THREAD

**Issue by chrishylton** [StackExchange/dapper-dot-net#394]
Curious if anyone has/is experienced any issues w/ Dapper (or any of the Dapper bolt on stuff like SimpleCRUD) following the install of .NET Framework 4.6.  Thinking out-loud that this could potentially be another JIT related issue w/ .NET 4.6, but not sure yet, still trying to figure out what in the world is going on.

Basically, I've got existing code that has always run fine (latest builds were under framework 4.5.1).  Once 4.6 got installed via group policy last week with a corporate update, performance went down the tubes.  I've got a series of .Query(Of MyThing)("select * from dbo.MyThing) statements, various objects/tables.  These tables span record counts of a few thousand up to around 1million rows per table (so some 250k, couple around 500k, couple at 1 million).  I'd say in total I'm loading 2.5-3 million objects, but don't quote me on that.  When it gets to the last table, only 4 columns, 250k records, the .Query call goes to lunch.  The entire process (all tables) takes 200 seconds, give or take.

Now the weird part.  If I move that last .Query up to the top and do it first, the entire process drops to 35 seconds.  That's the entire 2.5 millions rows loaded in 2 dozen calls to .Query.  Which is now back down to pre 4.6 performance.

I should note, this is all compiling as x64.  Performance is not an issue for x86 regardless of the order, other than depending on how much I load I can hit out-of-memory exceptions...but that's a different (expected) issue w/ running this much data under x86.

Any thoughts?

Thanks!
-Chris

**Comment by NickCraver** [comment]
There are 2 major changes in 4.6 that seem relevant: the JIT and GC behavior. Let's test and rule out GC first. Just for kicks try this `<runtime>` section if your `app.config` (including a full example for ease):
```XML
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
  <runtime>
    <gcServer enabled="true" />
    <gcAllowVeryLargeObjects enabled="true"/>
  </runtime>
  <startup>
    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5" />
  </startup>
</configuration>
```

**Comment by chrishylton** [comment]
Dude that looks to be it.  We've been down all the JIT rabbit holes we could find, including the great stuff you and/or Mark Gravell wrote about back in the summer.  But nothing had panned out, including the useLegacyJit switch.  I dropped those two settings into my test app here and I'm back to a 30-40 second run using the 'slower' version of the code.  Going to drop this into the production code, but the test there is a much longer running process that I'll have to confirm tomorrow.  But I think this does it.  Awesome!  Thanks for the help!  Was not aware of the GC side of the changes in .NET 4.6.

-Chris

**Comment by NickCraver** [comment]
Welcome! We actually hit this hard at Stack Overflow too - a nightly processes that processes 100-200 million rows of traffic logs each night went from minutes to hours. It sucks to hit, but I'm glad we could save you some time :) 

FWIW one of the GC regressions has a fix which I believe is in the 4.6.1 release, but it isn't RTM just yet.

**Comment by mgravell** [comment]
We should probably add an explicit "unexpectedly slow on 4.6?" page
somewhere. Hopefully phrased in a support way, not a "let's crap on 4.6"
way (we already did that once, cough JIT cough).

**Comment by chrishylton** [comment]
Yea, this wasn't nearly THAT big of a deal, but it was an annoyance that with this one app would have made it a pain to use once released.  This client actually has a second 'large data Dapper' app I wrote that will likely benefit from this tweak as well.  I jumped full steam into VS2015 when it officially dropped and had to make changes to this second app soon thereafter.  It ground to a halt, attempting to load largely the same data described above.  Didn't have the time to investigate at that point and ultimately removed VS2015, .NET 4.6, and moved on.  This time we had some time to investigate further and the app in question here isn't 'live' yet like the one a couple months back was.  So no complaint here, it is what it is.  Glad you guys knew right off what was up.
